include/binlog/Session.hpp

```diff
@@ -345,7 +347,8 @@ Session::ConsumeResult Session::consume(OutputStream& out)
     if (isClosed)
     {
       // queue is empty and closed, remove it
-      it = _channels.erase(it);
+      std::move(it+1, _channels.end(), it);
```

Comment:
Does the order of the vector matter? If not then you could swap the last element with the current and skip all the moves.

It also might be worth considering refactoring the loop to be separate erase and remove steps so removing all the elements are done in one go rather than one at a time?

Reply:
Order does not matter, but still wanted to keep the same behavior, to be less surprising in some cases. In the future, the vector might be replaced by a lockfree intrusive list (to save the lock in createChannel), that will have the same semantics.

Separate remove loop: not that trivial, as isClosed must be checked before the empty read, otherwise a race becomes possible:

 1. Consumer reads empty queue
 2. Producer adds data to the queue
 3. Producer closes the queue
 4. Consumer `std::remove_if`s closed queues - data added in step 2. is lost

To prevent this, the result of isClosed has to be stored in the first loop, not making the overall result much prettier.